src/box/alter.cc

```diff
@@ -3927,11 +3927,36 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
 					      BOX_SCHEMA_FIELD_KEY);
 	if (key == NULL)
 		return -1;
-	if (strcmp(key, "cluster") == 0) {
+	if (strcmp(key, "cluster") == 0 ||
+	    strcmp(key, "replicaset_uuid") == 0) {
+		/*
+		 * Ignore old 'cluster' if it was already replaced by the new
+		 * key 'replicaset_uuid'.
+		 */
+		char mpkey[64];
+		char *mpkey_end = mp_encode_array(mpkey, 1);
+		mpkey_end = mp_encode_str0(mpkey_end, "replicaset_uuid");
+		struct tuple *other = NULL;
+		if (box_index_get(BOX_SCHEMA_ID, 0, mpkey, mpkey_end,
+				  &other) != 0)
+			return -1;
+		bool is_new_version_stored = other != NULL;
 		if (new_tuple == NULL) {
-			diag_set(ClientError, ER_REPLICASET_UUID_IS_RO);
-			return -1;
+			/*
+			 * Don't allow to delete the old key until the new one
+			 * is here. Otherwise recovery wouldn't work in case
+			 * the instance goes down after old deletion but before
+			 * new insertion.
+			 */
+			if (strcmp(key, "replicaset_uuid") == 0 ||
+			    !is_new_version_stored) {
+				diag_set(ClientError, ER_REPLICASET_UUID_IS_RO);
+				return -1;
+			}
```

Comment:
Isn't the comment contradicts with the code? I mean that the comment is about the old key (`cluster`), while the code is about the new key (`replicaset_uuid`). Also, shouldn't it be tested?

And will not the code look more clean if we'll process the keys separately? There will be some duplication, but not much.

Reply:
I reworked this place from scratch and added new tests. Now I handle both keys the same way. The only restriction is that deletion of both of them should not possible. Otherwise recovery won't work.